cjworkbench/i18n/catalogs/extract.py

```diff
@@ -101,13 +101,75 @@ def extract_django(fileobj, keywords, comment_tags, options):
 
 
 def extract_python(fileobj, keywords, comment_tags, options):
+    def process(funcname, messages, translator_comments, message_lineno):
+        if funcname in ["trans", "trans_lazy"]:
+            # `messages` will have all the string parameters to our function
+            # As we specify in the documentation of `trans`,
+            # the first will be the message ID, the second will be the default message
+            # and the (optional) third will be the message context
+            if len(messages) > 1 and messages[1]:
+                # If we have a default, add it as a special comment
+                # that will be processed by our `merge_catalogs` script
+                translator_comments.append(
+                    (message_lineno, "default-message: " + messages[1])
+                )
+
+            if len(messages) > 2 and isinstance(messages[2], str):
+                context = messages[2]
+            else:
+                context = None
+
+            if context:
+                # if we have a context, trick pybabel to use `pgettext`
+                # so that it adds the context to the translation file
+                funcname = "pgettext"
+                messages = [context, messages[0]]
+            else:
+                funcname = None
+        return funcname, messages, translator_comments, message_lineno
+
+    return _parse_and_extract_python(fileobj, keywords, comment_tags, options, process)
+
+
+def extract_module_code(fileobj, keywords, comment_tags, options):
+    def process(funcname, messages, translator_comments, message_lineno):
+        if funcname == "trans":
```

Comment:
Isn't this what the `keywords` argument is for?

I would expect our signature to be: `def extract_module_code(fileobj): ...`

It would call `_parse_and_extract_python(fileobj, {"trans"}, {"i18n"})`

... then in this callback, it would be impossible for `funcname` to be anything other than `"trans"`, right?

Reply:
Regarding `funcname`, you are correct